planner/core/exhaust_physical_plans.go

```diff
@@ -537,78 +574,337 @@ func (p *LogicalJoin) constructInnerIndexScan(ds *DataSource, idx *model.IndexIn
+	rangeFilterCandidates := make([]expression.Expression, 0, len(innerPlan.pushedDownConds))
+	// This loop deal first filter out the expressions that contains columns not in index.
+	for _, innerFilter := range innerPlan.pushedDownConds {
+		affectedCols := expression.ExtractColumns(innerFilter)
+		if expression.ColumnSliceIsIntersect(affectedCols, ijHelper.curPossibleUsedKeys) {
+			remained = append(remained, innerFilter)
 			continue
 		}
-		usableKeys = append(usableKeys, key)
-		// Int datum 1 can convert to all column's type(numeric type, string type, json, time type, enum, set) safely.
-		fakeConstant := &expression.Constant{Value: types.NewIntDatum(1), RetType: key.GetType()}
-		eqFunc := expression.NewFunctionInternal(p.ctx, ast.EQ, types.NewFieldType(mysql.TypeTiny), key, fakeConstant)
-		conds = append(conds, eqFunc)
-		eqConds = append(eqConds, eqFunc)
+		rangeFilterCandidates = append(rangeFilterCandidates, innerFilter)
+	}
+	// Extract the eq/in functions of possible join key. This returned list keeps the same order with index column.
+	notKeyEqAndIn, remainedEqAndIn, rangeFilterCandidates, _ := ranger.ExtractEqAndInCondition(innerPlan.ctx, rangeFilterCandidates, ijHelper.curNotUsedIndexCols, ijHelper.curNotUsedColLens)
+	// We hope that the index cols appeared in the join keys can all be used to build range. If it cannot be satisfied,
+	// we'll mark this index as cannot be used for index join.
+	// So we should make sure that all columns before the keyMatchedLen is join key or has eq/in function.
+	if len(notKeyEqAndIn) < keyMatchedLen-matchedKeyCnt {
+		return nil, nil, nil
 	}
+	remained = append(remained, remainedEqAndIn...)
+	return notKeyEqAndIn, remained, rangeFilterCandidates
+}
 
-	// Look into every `innerFilter`, if it contains join keys' column, put this filter into `remained` part directly.
-	remained = make([]expression.Expression, 0, len(innerFilters))
-	for _, filter := range innerFilters {
-		affectedCols := expression.ExtractColumns(filter)
-		if expression.ColumnSliceIsIntersect(affectedCols, usableKeys) {
-			remained = append(remained, filter)
+func (ijHelper *indexJoinBuildHelper) analyzeOtherFilters(nextCol *expression.Column,
+	innerPlan *DataSource, cwc *ColWithCompareOps) []expression.Expression {
+	var lastColAccesses []expression.Expression
+loopOtherConds:
+	for _, filter := range ijHelper.join.OtherConditions {
+		sf, ok := filter.(*expression.ScalarFunction)
+		if !ok || !(sf.FuncName.L == ast.LE || sf.FuncName.L == ast.LT || sf.FuncName.L == ast.GE || sf.FuncName.L == ast.GT) {
 			continue
 		}
-		conds = append(conds, filter)
+		if lCol, ok := sf.GetArgs()[0].(*expression.Column); ok && lCol.Equal(nil, nextCol) {
+			affectedCols := expression.ExtractColumns(sf.GetArgs()[1])
+			if len(affectedCols) == 0 {
+				continue
+			}
+			for _, col := range affectedCols {
+				if innerPlan.schema.Contains(col) {
+					continue loopOtherConds
```

Comment:
Why we need it? Will it cause inf loops?

Reply:
No, the label just goto the next loop quickly. Since here it's a nested loop and we want to directly go to the next outer loop.